File: tools/sprite-onizle.py
#!/usr/bin/env python3
"""Sprite doğrulayıcı: satır uzunluklarını ve palet kapsamını kontrol eder,
her sprite'ı terminale basar. Yeni sprite çizdikten sonra çalıştır:

    python3 tools/sprite-onizle.py
"""
import re
import os
KOK = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
SRC = open(os.path.join(KOK, 'src/art/sprites.ts')).read()

# TS'teki P paletini ve her sprite'ın kendi palette override'ını gerçekten oku
pal_body = SRC.split('} as Record<string, string>;')[0].split('const P = {')[1]
BASE = set(re.findall(r"^\s*'?([A-Za-z.])'?:", pal_body, re.M))

blocks = re.findall(r'export const (\w+): SpriteDef = \{\n(.*?)\n\};', SRC, re.S)
bad = []
for name, body in blocks:
    head, rowspart = body.split('rows:')
    extra = set(re.findall(r"(\w+):\s*C\.", head))          # { ...P, h: C.rust }
    known = BASE | extra
    rows = re.findall(r"'([^']*)'", rowspart)
    lens = {len(r) for r in rows}
    unknown = sorted({c for r in rows for c in r if c not in known})
    status = []
    if len(lens) != 1: status.append(f"HIZASIZ {sorted(lens)}"); bad.append(name)
    if unknown:       status.append(f"PALETTE YOK: {unknown}"); bad.append(name)
    print(f"\n== {name}  {max(lens)}x{len(rows)}  {' | '.join(status) or 'ok'}")
    for i, r in enumerate(rows):
        mark = '<<' if len(r) != max(lens) else ''
        print('   ' + r.replace('.', ' ') + mark)
print(f"\n{'-'*30}\n{len(blocks)} sprite, sorunlu: {sorted(set(bad)) or 'yok'}")

File: tools/sprite-onizle.ts
/**
 * Sprite doğrulayıcı: satır uzunluklarını ve palet kapsamını kontrol eder,
 * istenirse her sprite'ı terminale basar. Yeni sprite çizdikten sonra:
 *
 *     npm run sprite            # yalnızca sorunlar ve özet
 *     npm run sprite -- --goster          # hepsini çiz
 *     npm run sprite -- --goster ASKER    # adında ASKER geçenleri çiz
 *
 * Eskiden Python'du ve sprites.ts'i regex'le okuyordu; `...ARKA_GOVDE` ya
 * da `rows: TISORT.rows` gibi başka tanımdan türeyen sprite'ları göremeyip
 * çöküyordu. Artık modülü doğrudan içe aktarıyor: yardımcı fonksiyonlarla
 * üretilen kareler de dahil, oyunun gördüğü veriyi doğruluyor.
 */
import * as S from '../src/art/sprites';
import type { SpriteDef } from '../src/ui/PixelSprite';

const argumanlar: string[] = process.argv.slice(2);
const goster = argumanlar.includes('--goster');
const suzgec = argumanlar.find((a) => !a.startsWith('--'))?.toUpperCase();

const spriteMi = (d: unknown): d is SpriteDef =>
  typeof d === 'object' &&
  d !== null &&
  Array.isArray((d as SpriteDef).rows) &&
  typeof (d as SpriteDef).palette === 'object';

/** Kare dizileri (giyinme, gece, sınav) tek tek doğrulanıyor. */
const tanimlar: [string, SpriteDef][] = [];
for (const [ad, deger] of Object.entries(S)) {
  if (spriteMi(deger)) tanimlar.push([ad, deger]);
  else if (Array.isArray(deger))
    deger.forEach((kare, i) => spriteMi(kare) && tanimlar.push([`${ad}[${i}]`, kare]));
}

const sorunlu: string[] = [];
for (const [ad, { palette, rows }] of tanimlar) {
  const durum: string[] = [];
  const uzunluklar = [...new Set(rows.map((r) => r.length))];
  const bilinmeyen = [...new Set(rows.join(''))].filter((c) => c !== '.' && !(c in palette)).sort();

  if (rows.length === 0) durum.push('BOŞ');
  if (uzunluklar.length > 1) durum.push(`HİZASIZ ${uzunluklar.sort((a, b) => a - b).join('/')}`);
  if (bilinmeyen.length) durum.push(`PALETTE YOK: ${bilinmeyen.join(' ')}`);
  if (durum.length) sorunlu.push(ad);

  const ciz = goster && (!suzgec || ad.includes(suzgec));
  if (!durum.length && !ciz) continue;

  const en = Math.max(0, ...uzunluklar);
  console.log(`\n== ${ad}  ${en}x${rows.length}  ${durum.join(' | ') || 'ok'}`);
  if (ciz || durum.length)
    for (const r of rows) console.log(`   ${r.replace(/\./g, ' ')}${r.length !== en ? ' <<' : ''}`);
}

console.log(`\n${'-'.repeat(30)}\n${tanimlar.length} sprite, sorunlu: ${sorunlu.length ? sorunlu.join(', ') : 'yok'}`);
process.exit(sorunlu.length ? 1 : 0);
